File: src/domain/releaseflow/ReleaseFlowRepository.ts
import { DataSource, EntityManager, FindOptionsWhere } from "typeorm";
import { FlowStatus, Stage } from "../../contracts/enums";
import { ReleaseFlowEntity } from "./ReleaseFlow.entity";

export interface ReleaseFlowFilter {
  /** Filter by project_id (from Excel "Project ID"). Maps to the ?project= query param. */
  projectId?: string;
  stage?: Stage;
  flowStatus?: FlowStatus;
}

export interface PageOptions {
  page: number;
  size: number;
}

export interface PageResult<T> {
  data: T[];
  total: number;
  page: number;
  size: number;
}

export class ReleaseFlowRepository {
  private readonly repo;

  constructor(ds: DataSource) {
    this.repo = ds.getRepository(ReleaseFlowEntity);
  }

  /** Use within an active transaction by passing a transactional EntityManager. */
  private repoFor(em?: EntityManager) {
    return em ? em.getRepository(ReleaseFlowEntity) : this.repo;
  }

  async findById(id: string, em?: EntityManager): Promise<ReleaseFlowEntity | null> {
    return this.repoFor(em).findOne({ where: { id }, relations: ["requests"] });
  }

  /**
   * Look up an existing Release Flow by grouping key: (projectId, normalizedReleaseId).
   * Used during import to determine whether to create a new RF or attach to an existing one.
   */
  async findByProjectIdAndNormalizedReleaseId(
    projectId: string,
    normalizedReleaseId: string,
    em?: EntityManager
  ): Promise<ReleaseFlowEntity | null> {
    return this.repoFor(em).findOne({
      where: { projectId, normalizedReleaseId },
    });
  }

  async findAll(
    filter: ReleaseFlowFilter,
    page: PageOptions
  ): Promise<PageResult<ReleaseFlowEntity>> {
    const where: FindOptionsWhere<ReleaseFlowEntity> = {};
    if (filter.projectId) where.projectId = filter.projectId;
    if (filter.flowStatus) where.flowStatus = filter.flowStatus;
    if (filter.stage) where.currentStage = filter.stage;

    const [data, total] = await this.repo.findAndCount({
      where,
      skip: page.page * page.size,
      take: page.size,
      order: { updatedAt: "DESC" },
    });
    return { data, total, page: page.page, size: page.size };
  }

  async save(entity: ReleaseFlowEntity, em?: EntityManager): Promise<ReleaseFlowEntity> {
    return this.repoFor(em).save(entity);
  }

  async updateStatus(
    id: string,
    patch: Partial<Pick<ReleaseFlowEntity, "flowStatus" | "reviewStatus" | "currentStage" | "reviewOwner">>,
    em?: EntityManager
  ): Promise<void> {
    await this.repoFor(em).update({ id }, patch);
  }
}

File: src/main/java/com/wwa/deploymentagent/domain/releaseflow/ReleaseFlowRepository.java
package com.wwa.deploymentagent.domain.releaseflow;

import com.wwa.deploymentagent.contracts.enums.FlowStatus;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;
import org.springframework.stereotype.Repository;

import java.util.Optional;
import java.util.List;

@Repository
public interface ReleaseFlowRepository extends JpaRepository<ReleaseFlow, String> {

    /**
     * Look up an existing Release Flow by grouping key: (projectId, normalizedReleaseId).
     * Used during import to determine whether to create a new RF or attach to existing one.
     */
    Optional<ReleaseFlow> findByProjectIdAndNormalizedReleaseIdAndArchivedAtIsNull(
            String projectId, String normalizedReleaseId);

    Optional<ReleaseFlow> findByProjectIdAndNormalizedReleaseId(
            String projectId, String normalizedReleaseId);

    @Query("""
            SELECT rf FROM ReleaseFlow rf
            WHERE (:projectId IS NULL OR rf.projectId = :projectId)
              AND (:flowStatus IS NULL OR rf.flowStatus = :flowStatus)
              AND (:stage IS NULL OR rf.currentStage = :stage)
              AND (:includeArchived = TRUE OR rf.archivedAt IS NULL)
            """)
    Page<ReleaseFlow> search(@Param("projectId") String projectId,
                             @Param("flowStatus") FlowStatus flowStatus,
                             @Param("stage") String stage,
                             @Param("includeArchived") boolean includeArchived,
                             Pageable pageable);

    /** Look up the first (active) Release Flow for a given projectId – used during import. */
    Optional<ReleaseFlow> findFirstByProjectIdAndArchivedAtIsNullOrderByCreatedAtDesc(String projectId);

    List<ReleaseFlow> findByProjectIdAndArchivedAtIsNullOrderByCreatedAtDesc(String projectId);

    @Query("""
            SELECT rf FROM ReleaseFlow rf
            WHERE rf.projectId = :projectId
              AND rf.archivedAt IS NULL
              AND NOT EXISTS (
                  SELECT 1 FROM Request r
                  WHERE r.releaseFlow = rf
                    AND r.stage = :stage
                    AND r.archivedAt IS NULL
              )
            ORDER BY rf.createdAt DESC
            """)
    List<ReleaseFlow> findActiveByProjectIdWithoutStageOrderByCreatedAtDesc(@Param("projectId") String projectId,
                                                                             @Param("stage") String stage);

    default Optional<ReleaseFlow> findFirstByProjectId(String projectId) {
        return findFirstByProjectIdAndArchivedAtIsNullOrderByCreatedAtDesc(projectId);
    }

    Optional<ReleaseFlow> findByIdAndArchivedAtIsNull(String id);

    /**
     * Load a Release Flow with its full request+task hierarchy in a single query.
     * Uses DISTINCT to avoid duplicates from multiple JOIN FETCH paths.
     */
    /** Count of existing Release Flows for a given projectId – used for seq generation. */
    @Query("SELECT COUNT(rf) FROM ReleaseFlow rf WHERE rf.projectId = :projectId")
    long countByProjectId(@Param("projectId") String projectId);
}
